File: code/875. Koko Eating Bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int H) {
        int n = piles.size();
        int r = *max_element(begin(piles), end(piles));
        int l = 1;
        int mid = -1;
        while (l <= r) {
            mid = l + (r - l) / 2;
            int cnt = 0;
            for (int& p : piles) {
                cnt = cnt + p / mid + (p % mid == 0 ? 0 : 1);
            }
            if (l == r) break;
            if (cnt <= H) r = mid;
            else l = mid + 1;
        }
        return mid;
    }
};

File: code/875. Koko Eating Bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], H: int) -> int:
        l, r = 1, max(piles)
        k = 0
        while l <= r:
            k = (l + r) // 2
            if l == r:
                break
            hours = 0
            for p in piles:
                hours += p // k
                if p % k:
                    hours += 1
            if hours <= H:
                r = k
            else:
                l = k + 1
        return k
            